teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,494 @@
+      results:
+        - name: ARTIFACT-INPUTS_git-vcs
+          value:
+            uri: git+https://github.com/foo/bar.git
+            digest: sha256:abc
+        - name: ARTIFACT-OUTPUTS_oci-image
+          value:
+            uri: gcr.io/somerepo/someimage
+            digest: sha512:abc
+    ```
+The generated intoto provenance will contain the following:
+```
+subjects: [{"name": "gcr.io/somerepo/someimage", "digest": {"sha256": "abc"}}]
+...
+materials: [{
+  "uri": "git+https://github.com/foo/bar.git",
+  "digest": {"sha1": "abc..."}
+}]
+```
+
+2. Use Perforce and images as sources, and generates in-toto provenance for a TaskRun that builds a Maven package.
+``` yaml
+results:
+  - name: ARTIFACT-INPUTS_perforce-vcs
+    value:
+      uri: http://myp4web:8080/depot/main/atlas/
+      digest: sha256:abc
+  - name: ARTIFACT-OUTPUTS_maven-pkg
+    value:
+      uri: us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.jar
+      digest: sha256:abc
+  - name: ARTIFACT-OUTPUTS_maven-pom
+    value:
+      uri: us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.pom
+      digest: sha256:def
+  - name: ARTIFACT-OUTPUTS_maven-src-pkg
+    value:
+      uri: us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0-sources.jar
+      digest: sha256:xyz
+```
+The generated field for these targets will be:
+```
+subjects: [{"name": "us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.jar", "digest": {"sha256": "abc"}}
+{"name": "us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0.pom", "digest": {"sha256": "def"}}
+{"name": "us-west4-maven.pkg.dev/test-project/test-repo/com/google/guava/guava/31.0/guava-31.0-sources.jar", "digest": {"sha256": "xyz"}}]
+...
+materials: [{
+  // The git repo that contains the build.yaml referenced above.
+  "uri": "http://myp4web:8080/depot/main/atlas/",
+  // The resolved git commit hash reflecting the version of the repo used
+  // for this build.
+  "digest": {"sha1": "abc..."}
+}]
+```
+
+### Requirements
+
+- Able to structurally retrieve provenance info from TaskRun and produces currently supported in-toto provenance formats once a TaskRun is finished
+- Based on SLSA L2 requirements, the provenance must provide the location of source code in version control and the provenance MUST identify the output artifact via at least one cryptographic hash.
+- If a TaskRun is within a PipelineRun, the TaskRun’s provenance can be produced between the time that when one TaskRun is finished
+
+## Proposal
+
+**_Please note that the below proposal will be acted as a middle ground proposal, as it on one side definitely provides more organized structure for users to provide provenance data in Tekton TaskRuns and PipelineRuns, but on the other side that it still has a bunch of trade offs that can be resolved by another proposal mentioned in Alternatives section._**
```

Comment:
It might help to refocus this TEP on the desired end state, and move the middle ground impl to `Implementation Plan` and/or `Risks and Mitigations`.

We should also look out to fill out the sections for `Drawbacks` - though I think this should focus on the end goal rather than in-between steps.